File: src/mem/mtree/timing_tree.cc
#include "mem/mtree/timing_tree.hh"

#include <iomanip>

#include "base/trace.hh"
#include "debug/TimingTree.hh"
#include "debug/TimingTreeInit.hh"
#include "mem/mtree/util.hh"

namespace gem5 {

TimingTree::TimingTree(unsigned int arity, uint64_t total_data) :
  arity(arity),
  hashInputSize(BLOCK_SIZE_BYTES),
  hashOutputSize(hashInputSize / arity)
{
  // Note this should be in sync with the Python version

  assert(arity > 0);

  // Calculate the number of leaves needed to get this much data

  // Consider the total number of hashes necessary to cover this area.
  long long leafHashes = total_data / hashInputSize;
  // Handle possible extra remainder.
  if (total_data % hashInputSize != 0) leafHashes++;
  DPRINTF(TimingTreeInit, "%s: %lld total leaf hashes needed.\n", __func__,
      leafHashes);

  long long lastLevelNodes = leafHashes / arity;
  // Handle possible extra remainder.
  if (leafHashes % arity != 0) lastLevelNodes++;
  DPRINTF(TimingTreeInit, "%s: %lld last level nodes.\n",
    __func__, lastLevelNodes);
  leaves = lastLevelNodes;

  // Calculate how many levels you need to get that many leaves in the tree.
  unsigned int levels = integerLog(lastLevelNodes, arity) + 1;
  height = levels;
  DPRINTF(TimingTreeInit, "%s: The height is %u.\n", __func__, levels);

  // Initialize the data size parameter. Take the number of nodes for all the
  // fully-filled levels, then add the number of leaf nodes.
  dataSize = integerPower(arity, levels - 1)/(arity - 1);
  dataSize += lastLevelNodes;
  DPRINTF(TimingTreeInit, "%s: Total number of nodes is %u.\n",
    __func__, dataSize);

  DPRINTF(TimingTreeInit, "%s: Total space protected by tree: %llu bytes. "
    "(Requested: %llu bytes.)\n",
    __func__, statDataProtected(), total_data);

  DPRINTF(TimingTreeInit, "%s: Total space taken by tree: %llu bytes.\n",
    __func__, statStructureSize());
}

TimingTree::~TimingTree() {
}

void TimingTree::processWrite(
  size_t address,
  std::vector<unsigned char> newData
) {
  std::cout << "====================================" << std::endl;
  std::cout << "Processing write at address " << address << "." << std::endl;
  std::cout << "Input data: ";
  printHex(&newData[0], newData.size());

  // Time to start hashing!
  // Update the hash for the leaf that directly corresponds to this address.
  updateLeafHash(address);

  // Update up the rest of the tree.
  size_t childIndex = addressToBlockIndex(address);
  size_t parentIndex = parentBlockIndex(childIndex);

  while (true) {
    try {
      // Find exactly where the block is relative to its parent.
      size_t relativeChildIndex = relativeChildBlockIndex(childIndex);

      // Partially update the parent block based on this child.
      updateBlockHashPartial(parentIndex, relativeChildIndex);

      // Ascend the tree up one level.
      childIndex = parentIndex;
      parentIndex = parentBlockIndex(parentIndex);
    }
    catch (int exception) {
      // No more parents. Stop ascending the tree.
      break;
    }
  }

  // Update top node (leads to secure root).
  // (input one node, output a single hash)
}


unsigned int TimingTree::processWriteTiming(size_t address) {
  // Keep track of the amount of time required to complete write (in cycles).
  unsigned int totalTime = 0;

  std::cout << "====================================" << std::endl;
  std::cout << "Processing write at address " << address << "." << std::endl;

  // Time to start hashing!
  // Update the hash for the leaf that directly corresponds to this address.
  totalTime += updateLeafHash(address);

  // Update up the rest of the tree.
  size_t childIndex = addressToBlockIndex(address);
  size_t parentIndex = parentBlockIndex(childIndex);

  while (true) {
    try {
      // Find exactly where the block is relative to its parent.
      size_t relativeChildIndex = relativeChildBlockIndex(childIndex);

      // Partially update the parent block based on this child.
      totalTime += updateBlockHashPartial(parentIndex, relativeChildIndex);

      // Ascend the tree up one level.
      childIndex = parentIndex;
      parentIndex = parentBlockIndex(parentIndex);
    }
    catch (int exception) {
      // No more parents. Stop ascending the tree.
      break;
    }
  }

  // Update top node (leads to secure root).
  // (input one node, output a single hash)

  return totalTime;
}


size_t TimingTree::parentBlockIndex(size_t index) {
  if (index == 0) {
    // There is no parent.
    throw -1;
  }

  return (index - 1) / arity;
}


size_t TimingTree::firstChildBlockIndex(size_t index) {
  return (arity * index) + 1;
}


size_t TimingTree::relativeChildBlockIndex(size_t childIndex) {
  // Get the parent index, find its first child, then get the difference
  // between the requested child and the first child.

  size_t parentIndex = parentBlockIndex(childIndex);
  size_t firstChildIndex = firstChildBlockIndex(parentIndex);

  return childIndex - firstChildIndex;
}


bool TimingTree::isLeaf(size_t index) {
  size_t firstLeafIndex = integerPower(arity, height - 1)/(arity - 1);
  return (index >= firstLeafIndex);
}


bool TimingTree::isAncestor(size_t parent, size_t child) {
  assert(parent != child);

  while (parent < child) {
    child = parentBlockIndex(child);
    if (child < parent) {
      // Child hopped to the same level as or above the parent.
      // Not an ancestor.
      return false;
    } else if (child == parent) {
      // This is an ancestor.
      return true;
    }
  }

  return false;
}

AbstractIntegrityTree::TreeNodeType
TimingTree::getNodeType(size_t index) {
  // All nodes in the TimingTree are regular tree nodes.
  return TreeNodeType::TreeNode;
}


size_t TimingTree::addressToBlockIndex(size_t address) {
  // Get the index of the first leaf node.
  size_t firstLeafIndex = integerPower(arity, height - 1)/(arity - 1);
  // DPRINTF(TimingTree, "%s: We have %d total nodes.\n", __func__, dataSize);
  // DPRINTF(TimingTree, "%s: The first leaf is at index %d.\n",
        // __func__, firstLeafIndex);
  // DPRINTF(TimingTree, "%s: The arity is %d.\n", __func__, arity);

  // unsigned int nodesOnLevel = 1;
  // unsigned int firstNodeIndex = 0;
  // size_t i = 0;

  // for (unsigned int level = 0; level < height; level++) {

  //   std::cout << "There are "
  //        << nodesOnLevel << " nodes on level " << level << std::endl;
  //   std::cout << "The first index of this level is "
  //        << firstNodeIndex << std::endl;

  //   firstNodeIndex += nodesOnLevel;
  //   nodesOnLevel *= arity;
  // }

  // Adjust the address to ensure it is aligned to the cache line size.
  address = address & ~(BLOCK_SIZE_BYTES - 1);

  // Offset this index by the amount of memory covered up to this address.

  // We have the first leaf, so now we need to calculate the offset.
  // One leaf block covers `inputHashSize * arity` bytes.
  // So take the address and divide by this amount.
  size_t leafOffset = address / (hashInputSize * arity);
  // DPRINTF(TimingTree,
      // "%s: The leaf that corresponds to this address is #%d\n",
      // __func__, leafOffset);

  return firstLeafIndex + leafOffset;
}



uint64_t TimingTree::simulatedBlockOffset(size_t index) {
  return (index * BLOCK_SIZE_BYTES);
}


unsigned int TimingTree::updateLeafHash(size_t address) {
  std::cout << "====================================" << std::endl;
  std::cout << "Processing write at address " << address << "." << std::endl;

  // Calculate which leaf corresponds to this request's address.
  [[maybe_unused]] size_t leafIndex = addressToBlockIndex(address);

  // Find exactly which portion of this block is relevant to this address.
  size_t blockOffset = (address / hashInputSize) % arity;
  std::cout << "The portion of the leaf for this is hash #"
    << blockOffset << std::endl;

  // Time to start hashing!
  // Update the hash for the leaf that directly corresponds to this address.
  // outputHash = hash(newData)

  // leaf between [`byteOffset` and `byteOffset + hashOutputSize`) = outputHash

  // total time = identify block + load + hash + store
  return CYCLES_READ_DATA + CYCLES_COMPUTE_HASH + CYCLES_WRITE_DATA;
}


void TimingTree::updateBlockHash(size_t index) {
  for (size_t i = 0; i < arity; i++) {
    updateBlockHashPartial(index, i);
  }
}


unsigned int TimingTree::updateBlockHashPartial(
  size_t index,
  size_t childIndex
) {
  // The block being updated.
  std::cout << "Updating block " << index << "." << std::endl;

  // The child block that has changed and needs to propagate its hash up.
  [[maybe_unused]] size_t childBlock = firstChildBlockIndex(index)
                                        + childIndex;

  // Determine where to start changing data in the updated block.
  size_t byteOffset = childIndex * hashOutputSize;

  std::cout << "Updating bytes #" << byteOffset << " through #"
    << (byteOffset + hashOutputSize - 1) << "." << std::endl;

  // Time to start hashing!
  // Get the child block's data and hash it.
  // hash(childBlock)

  // Copy the hash to the parent block.
  // block[byteOffset] = hash(childBlock)
  return CYCLES_READ_DATA + CYCLES_COMPUTE_HASH + CYCLES_WRITE_DATA;
}


long long TimingTree::statDataProtected() {
  // NOTE Sync with Python

  // Assuming that in the basic version, each leaf contains `arity` number of
  // hashes for real data.

  // Consider the total number of hashes created by the set of leaves.
  long long leafHashes = leaves * arity;

  // Consider the amount of data represented by a single hash.
  return leafHashes * hashInputSize;
}

long long TimingTree::statStructureSize() {
  // NOTE Sync with Python

  return dataSize * BLOCK_SIZE_BYTES;
}

double TimingTree::statDataOverheadRatio() {
  return (double) statStructureSize() / statDataProtected();
}

} // namespace gem5

File: src/mem/mtree/TimingTree.py
# Only used to essentially provide the Python version of a C++ function

from math import (
    ceil,
    log,
)
from typing import Tuple


def truncate_to_size_multiple(original_size: int, size: int) -> int:
    # Truncate protected local size to a multiple of a size.
    remainder = original_size % size
    if remainder != 0:
        original_size -= remainder

    return original_size


class TimingTree:
    def __init__(
        self,
        arity: int,
        total_protected_data: int,
    ) -> None:
        self.arity = arity
        self.total_protected_data = total_protected_data

        # Note that this should be in sync with the C++ version

        # Consider the total number of hashes necessary to cover this area.
        leaf_hashes = int(
            self.total_protected_data / self._get_hash_input_size()
        )
        # Handle possible extra remainder.
        if self.total_protected_data % self._get_hash_input_size() != 0:
            leaf_hashes += 1
        # print(f"PYTHON: {leaf_hashes} total leaf hashes needed.")

        last_level_nodes = int(leaf_hashes / self.arity)
        # Handle possible extra remainder.
        if leaf_hashes % self.arity != 0:
            last_level_nodes += 1
        # print(f"PYTHON: {last_level_nodes} last level nodes.")
        self.leaves = last_level_nodes

        # Calculate how many levels you need to get that many leaves in the tree.
        levels = ceil(log(last_level_nodes, self.arity)) + 1
        self.height = levels
        # print(f"PYTHON: The height is {levels}.")

        # Initialize the data size parameter. Take the number of nodes for all the
        # fully-filled levels, then add the number of leaf nodes.
        self.data_size = int((self.arity ** (levels - 1)) / (self.arity - 1))
        self.data_size += last_level_nodes
        # print(f"PYTHON: Total number of nodes is {self.data_size}.")

        # print(
        #     f"PYTHON: Total space protected by tree: {self._stat_data_protected()} bytes. (Requested: {self.total_protected_data})"
        # )

        # print(
        #     f"PYTHON: Total space taken by tree: {self._stat_structure_size()} bytes."
        # )

    def _get_block_size_bytes(self) -> int:
        return 64

    def _get_hash_input_size(self) -> int:
        return self._get_block_size_bytes()

    def _get_hash_output_size(self) -> int:
        return int(self._get_hash_input_size() / self.arity)

    def _stat_data_protected(self) -> int:
        # NOTE Sync with C++

        # Assuming that in the basic version, each leaf contains `arity` number
        # of hashes for real data.

        # Consider the total number of hashes created by the set of leaves.
        leaf_hashes = self.leaves * self.arity

        # Consider the amount of data represented by a single hash.
        return int(leaf_hashes * self._get_hash_input_size())

    def _stat_structure_size(self) -> int:
        # NOTE Sync with C++

        return int(self.data_size * self._get_block_size_bytes())

    @staticmethod
    def determine_max_protected_size(
        min_local_size: int,
        max_local_size: int,
        total_local_size: int,
        total_remote_size: int,
        arity: int,
    ) -> Tuple[int, int]:
        """
        :param min_local_size:    Minimum amount of size that must be protected
                                  in local memory.
        :param max_local_size:    Maximum amount of size that must be protected
                                  in local memory. If this option is `0`, the
                                  theoretical maximum is `total_local_size`.
        :param total_local_size:  Total amount of memory available to work with
                                  in local memory.
        :param total_remote_size: Total amount of memory available to work
                                  with in remote (essentially static
                                  contributor to available space).
        :param arity:             The arity of the tree structure used.

        Returns the amount of data at most (in local, and remote memory) that
        can be protected based on the amount of data it would take to protect
        it, and the total amount of space available.

        Two types of memory: local and remote. ("Local" in this case means
        "the same place as integrity data." "Remote" means "the other
        place that is not the same place as integrity data.")

        This is not an actual bearing on the memory architecture used.

        NOTE: Currently assumes only one type of memory being used to store
              integrity data at a time.
        """

        # Perform a binary search to find the max amount of data to protect

        # Scale factor represents the amount of data to use in local memory.
        if min_local_size != 0:
            min_scale_factor = 1
            if max_local_size != 0:
                max_scale_factor = max_local_size / min_local_size
            else:
                max_scale_factor = total_local_size / min_local_size
        else:
            min_scale_factor = 0
            max_scale_factor = 0
        selected_scale_factor = min_scale_factor + (
            (max_scale_factor - min_scale_factor) / 2
        )
        cache_line_size = 64
        page_size = 4096

        assert total_remote_size % cache_line_size == 0

        # print(
        #     f"TimingTree.determine_max_protected_size: min_local_size: {min_local_size}"
        # )
        # print(
        #     f"TimingTree.determine_max_protected_size: total_local_size: {total_local_size}"
        # )
        # print(
        #     f"TimingTree.determine_max_protected_size: total_remote_size: {total_remote_size}"
        # )
        # print(
        #     f"TimingTree.determine_max_protected_size: min_scale_factor: {min_scale_factor}"
        # )
        # print(
        #     f"TimingTree.determine_max_protected_size: max_scale_factor: {max_scale_factor}"
        # )

        iterations = 0
        while True:
            min_local_protected = truncate_to_size_multiple(
                int(min_local_size * min_scale_factor), cache_line_size
            )
            max_local_protected = truncate_to_size_multiple(
                int(min_local_size * max_scale_factor), cache_line_size
            )

            attempted_local_protected = int(
                min_local_size * selected_scale_factor
            )
            attempted_local_protected = truncate_to_size_multiple(
                attempted_local_protected, cache_line_size
            )

            # Calculate the amount of space we may end up protecting
            attempted_protected_size = (
                total_remote_size + attempted_local_protected
            )

            # (For integrity data)
            remaining_local_space = (
                total_local_size - attempted_local_protected
            )

            # Calculate tree size
            # print(f"TimingTree.determine_max_protected_size: ==============")
            tree = TimingTree(arity, attempted_protected_size)

            # print(
            #     f"TimingTree.determine_max_protected_size: min_scale_factor: {min_scale_factor} (Local = {min_local_protected}, Remote = {total_remote_size}, Total = {min_local_protected + total_remote_size})"
            # )
            # print(
            #     f"TimingTree.determine_max_protected_size: selected_scale_factor: {selected_scale_factor} (Local = {attempted_local_protected}, Remote = {total_remote_size}, Total = {attempted_protected_size})"
            # )
            # print(
            #     f"TimingTree.determine_max_protected_size: --> Available space: {remaining_local_space}"
            # )
            # print(
            #     f"TimingTree.determine_max_protected_size: --> Tree size: {tree._stat_structure_size()}"
            # )
            # print(
            #     f"TimingTree.determine_max_protected_size: --> Difference: {remaining_local_space - tree._stat_structure_size()}"
            # )
            # print(
            #     f"TimingTree.determine_max_protected_size: max_scale_factor: {max_scale_factor} (Local = {max_local_protected}, Remote = {total_remote_size}, Total = {max_local_protected + total_remote_size})"
            # )

            if (
                remaining_local_space - tree._stat_structure_size() >= 0
                and remaining_local_space - tree._stat_structure_size()
                < page_size
            ):
                # We have found the best arrangement of tree size with this amount of protected space. Stop.
                attempted_local_protected = truncate_to_size_multiple(
                    attempted_local_protected, page_size
                )
                print(
                    f"TimingTree.determine_max_protected_size: Scale factor found."
                )
                return attempted_local_protected, total_remote_size
            elif tree._stat_structure_size() < remaining_local_space:
                # There is more space that could be protected, since there is leftover space on the table.
                # print(
                #     f"TimingTree.determine_max_protected_size: Increasing scale factor."
                # )
                min_scale_factor = selected_scale_factor
            else:
                # This structure is too large for the amount of space that is leftover, reduce the protected size to make more space for integrity data
                # print(
                #     f"TimingTree.determine_max_protected_size: Decreasing scale factor."
                # )
                max_scale_factor = selected_scale_factor

            selected_scale_factor = min_scale_factor + (
                (max_scale_factor - min_scale_factor) / 2
            )
            iterations += 1

            if iterations > 500:
                if (
                    remaining_local_space - tree._stat_structure_size()
                    >= page_size
                ):
                    print(
                        "It looks like this configuration is leaving unused space, but this is still a valid configuration."
                    )
                    return attempted_local_protected, total_remote_size

                print(
                    "Could not find a valid size arrangement for this configuration. (The integrity tree may take up too much space no matter what.)"
                )
                exit(1)
